controlplane/src/core/services/Keycloak.ts

```diff
@@ -115,6 +115,104 @@ export default class Keycloak {
     });
   }
 
+  public async getKeycloakUsers({ realm, orgSlug }: { realm?: string; orgSlug: string }) {
+    const users = await this.client.users.find({
+      exact: true,
+      idpAlias: orgSlug,
+      realm: realm || this.realm,
+    });
+    return users;
+  }
+
+  public async getKeycloakUserGroups({ realm, userID }: { realm?: string; userID: string }) {
+    const groups = await this.client.users.listGroups({
+      id: userID,
+      realm: realm || this.realm,
+    });
+    return groups;
+  }
+
+  public async createOIDCProvider({
+    realm,
+    clientId,
+    clientSecret,
+    name,
+    orgSlug,
+    discoveryEndpoint,
+  }: {
+    realm?: string;
+    clientId: string;
+    clientSecret: string;
+    name: string;
+    orgSlug: string;
+    discoveryEndpoint: string;
+  }) {
+    const oidcUrls = await this.client.identityProviders.importFromUrl({
+      realm: realm || this.realm,
+      fromUrl: discoveryEndpoint,
+      providerId: 'oidc',
+    });
+
+    await this.client.identityProviders.create({
+      alias: orgSlug,
+      displayName: name,
+      enabled: true,
+      config: {
+        clientId,
+        clientSecret,
+        hideOnLoginPage: true,
+        syncMode: 'FORCE',
+        validateSignature: 'true',
+        tokenUrl: oidcUrls.tokenUrl,
+        authorizationUrl: oidcUrls.authorizationUrl,
+        jwksUrl: oidcUrls.jwksUrl,
+        logoutUrl: oidcUrls.logoutUrl,
+        issuer: oidcUrls.issuer,
+        useJwksUrl: 'true',
+        defaultScope: 'openid email',
+      },
+      realm: realm || this.realm,
+      providerId: 'oidc',
+    });
+  }
+
+  public async deleteOIDCProvider({ realm, orgSlug }: { realm?: string; orgSlug: string }) {
+    await this.client.identityProviders.del({ alias: orgSlug, realm: realm || this.realm });
+  }
+
+  public async createIDPMapper({
+    realm,
+    claims,
+    orgSlug,
+    keycloakGroupName,
+  }: {
+    realm?: string;
+    claims: string;
+    orgSlug: string;
+    keycloakGroupName: string;
+  }) {
+    try {
+      const a = await this.client.identityProviders.createMapper({
+        alias: orgSlug,
+        realm: realm || this.realm,
+        identityProviderMapper: {
+          name: uid(10),
+          identityProviderMapper: 'oidc-advanced-group-idp-mapper',
+          identityProviderAlias: orgSlug,
+          config: {
+            claims,
+            syncMode: 'INHERIT',
+            group: keycloakGroupName,
+            'are.claim.values.regex': true,
```

Comment:
How do we catch invalid regex? What's the supported syntax? Did you communicate this feature on creation? Is it worth to enable it in the first version?

Reply:
keycloak just takes a string, if the regex is right it will work or else the mapping won't work.
Yes, we tell the users to either give a regex or the group name.
I thought there shouldn't be any problems even if we take regex

Reply:
one thing is that, it might be difficult for the users to debug if the mapping doesn't work as the regex is wrong

Reply:
That's fine for me.